storage/engine: move MVCCIncrementalIterator from storageccl to engine

hueypark opened the pull request:
This is prerequisites for [storage/engine: add Pebble support to ExportToSst](https://github.com/cockroachdb/cockroach/issues/41739). Recovered some tests from [engine: Add ExportToSst method with all logic moved to C++](https://github.com/cockroachdb/cockroach/commit/75f1314b9ad4326069784838c581e1cb7a5ede24).

See #41945

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41991)
<!-- Reviewable:end -->

petermattis reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark)

___
*[pkg/storage/engine/mvcc_incremental_iterator_test.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41991#-LsROHqJ9RraV7_2Gtmb:-LsROHqJ9RraV7_2Gtmc:b-45idpu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9ef0d80b6ce71c240efef5542d9a93adebac62f/pkg/storage/engine/mvcc_incremental_iterator_test.go#L95)):*
> ```Go
> }
> 
> func TestMVCCIterateIncremental(t *testing.T) {
> ```

Did these tests come from somewhere or are they brand new? I need to scrutinize them more. My first reaction is "quite nice", though for many of the MVCC layer tests I'd like to be pushing towards using the `datadriven` framework. That isn't necessary in this PR, I'm mostly just curious about their pedigree. 
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41991#-LsN6Xur48rfX14KQ5pV:-LsN6Xur48rfX14KQ5pW:b-7rczkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9ef0d80b6ce71c240efef5542d9a93adebac62f/pkg/storage/engine/mvcc_incremental_iterator.go#L53)):*
> ```Go
> // oracle to prove the correctness of the new export logic.
> type MVCCIncrementalIterator struct {
> 	// TODO(dan): Move all this logic into c++ and make this a thin wrapper.
> ```

You can remove this TODO. There is a C++ version of all of this logic.s


<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis)

___
*[pkg/storage/engine/mvcc_incremental_iterator_test.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41991#-LsROHqJ9RraV7_2Gtmb:-LsRiXID1hIDnd0xd2mk:b-bfoz5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9ef0d80b6ce71c240efef5542d9a93adebac62f/pkg/storage/engine/mvcc_incremental_iterator_test.go#L95)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did these tests come from somewhere or are they brand new? I need to scrutinize them more. My first reaction is "quite nice", though for many of the MVCC layer tests I'd like to be pushing towards using the `datadriven` framework. That isn't necessary in this PR, I'm mostly just curious about their pedigree. 
</blockquote></details>

It' deleted tests from [engine: Add ExportToSst method with all logic moved to C++](https://github.com/cockroachdb/cockroach/commit/75f1314b9ad4326069784838c581e1cb7a5ede24).
___
*[pkg/storage/engine/mvcc_incremental_iterator.go, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41991#-LsN6Xur48rfX14KQ5pV:-LsRhy_r7kxbawAmAyJV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9ef0d80b6ce71c240efef5542d9a93adebac62f/pkg/storage/engine/mvcc_incremental_iterator.go#L53)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You can remove this TODO. There is a C++ version of all of this logic.s
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo a few small comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41991#-:-LsRwGc7DHwSNBK7VDe4:bc4c7z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @hueypark)

___
*[pkg/storage/engine/mvcc_incremental_iterator_test.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41991#-LsROHqJ9RraV7_2Gtmb:-LsRv7A21wyVFW10278G:bt7md4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9ef0d80b6ce71c240efef5542d9a93adebac62f/pkg/storage/engine/mvcc_incremental_iterator_test.go#L95)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

It' deleted tests from [engine: Add ExportToSst method with all logic moved to C++](https://github.com/cockroachdb/cockroach/commit/75f1314b9ad4326069784838c581e1cb7a5ede24).
</blockquote></details>

Ack.

Nit: `s/TestMVCCIterateIncremental/MVCCIncrementalIterator/g` for mildly improved grep'ability of this code.
___
*[pkg/storage/engine/mvcc_incremental_iterator_test.go, line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41991#-LsRvTssAsrGsFJkwalB:-LsRvTstAx656c5ZjNtm:b-h3fdcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba4db5e46eaa8c0e1fa72d91fc69861c17f76c26/pkg/storage/engine/mvcc_incremental_iterator_test.go#L129)):*
> ```Go
> 	kvs := func(kvs ...MVCCKeyValue) []MVCCKeyValue { return kvs }
> 
> 	runWithAllEngines(func(e Engine, t *testing.T) {
> ```

I have a TODO to replace `runWithAllEngines` with:

```
	for _, engineImpl := range mvccEngineImpls {
		t.Run(engineImpl.name, func(t *testing.T) {
			engine := engineImpl.create()
			defer engine.Close()
```

This allows us to select the subtests from the command line. We could technically fix `runWithAllEngines` to use subtests, but I'd like to have one mechanism, and the for-loop is used more widely.


<!-- Sent from Reviewable.io -->

petermattis commented:
@danhhz Note that the C++ version of incremental iterator was already in non-ccl code. That might have been an error when that code was written, but I don't think this changes the situation materially.

petermattis commented:
@danhhz Do you want `ExportToSst` to exist in CCL code? Currently it exists in non-CCL C++ code.

danhhz commented:
I would have put ExportToSst and DBExportToSst in ccl, but at this point our hands are tied, no?

My bigger worry is that it ended up out of ccl in the first place and that's now having a cascading effect. Are we worried that's going to happen again?

petermattis commented:
> My bigger worry is that it ended up out of ccl in the first place and that's now having a cascading effect. Are we worried that's going to happen again?

Yeah. Do you know what happened there? I don't.

danhhz commented:
Looks like it happened in https://github.com/cockroachdb/cockroach/pull/37496. I peeked and as of this PR, MVCCIncrementalIterator was still in ccl (pkg/ccl/storageccl/engineccl/mvcc.go), so not sure why the c++ port of it ended up outside ccl. The right people (david) were on the review. I don't see any discussion in the review about ccl.

hueypark committed:
storage/engine: move MVCCIncrementalIterator from storageccl to engine

This is prerequisites for [storage/engine: add Pebble support to ExportToSst](https://github.com/cockroachdb/cockroach/issues/41739). Recovered some tests from [engine: Add ExportToSst method with all logic moved to C++](https://github.com/cockroachdb/cockroach/commit/75f1314b9ad4326069784838c581e1cb7a5ede24).

See #41945

Release note: None

hueypark reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41991)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @petermattis)

___
*[pkg/storage/engine/mvcc_incremental_iterator_test.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41991#-LsROHqJ9RraV7_2Gtmb:-LsWeO-q-_PgVBv-T15k:b3oh8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9ef0d80b6ce71c240efef5542d9a93adebac62f/pkg/storage/engine/mvcc_incremental_iterator_test.go#L95)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ack.

Nit: `s/TestMVCCIterateIncremental/MVCCIncrementalIterator/g` for mildly improved grep'ability of this code.
</blockquote></details>


Done.
___
*[pkg/storage/engine/mvcc_incremental_iterator_test.go, line 129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41991#-LsRvTssAsrGsFJkwalB:-LsWh3ar7JS4RfCgnBrI:b3qvewm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ba4db5e46eaa8c0e1fa72d91fc69861c17f76c26/pkg/storage/engine/mvcc_incremental_iterator_test.go#L129)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I have a TODO to replace `runWithAllEngines` with:

```
	for _, engineImpl := range mvccEngineImpls {
		t.Run(engineImpl.name, func(t *testing.T) {
			engine := engineImpl.create()
			defer engine.Close()
```

This allows us to select the subtests from the command line. We could technically fix `runWithAllEngines` to use subtests, but I'd like to have one mechanism, and the for-loop is used more widely.
</blockquote></details>

I created an issue for this: https://github.com/cockroachdb/cockroach/issues/42060


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
@hueypark the change here looks good, but we're still figuring out code structuring. 

@danhhz So what should we do here? Should we keep the Go version of incremental iterator in CCL code? 

Cc @dt in case he has opinions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41991#-:-LsXCGv604l65VkJa3zu:b-a6uf5v)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

danhhz commented:
> @danhhz So what should we do here? Should we keep the Go version of incremental iterator in CCL code?

Do you see a good way to keep the go incremental iterator in ccl? I do not. To use the hook method, we'd have to create some interface in non-ccl code that it implements and write ExportToSst in terms of it. I doubt the performance here would be good. I see 3 options:

1) Move go incremental iterator out of ccl. A lot of export/import has been moving out of ccl anyway (see ExportStorage), so maybe we've lost this battle. Learn our lesson and figure out if there's something additional we should be doing to prevent this from happening to changefeeds, etc that are still firmly in ccl.
2) ExportToSst was added in the 19.2 cycle. We've previously been okay with moving stuff into ccl when it was merged apache2 by mistake as long as it hasn't made it into a major release. So, sneak a PR that moves ExportToSst to ccl into 19.2.0.
3) Same as 2 but 19.2.1.

I'm happy with 1 or 3. Thoughts? Is there an option I'm missing?

> Cc @dt in case he has opinions.

He's OOO, but calendar indicates he's back tomorrow? If possible, I would like his opinion here before we commit to anything.

petermattis commented:
Hmm, I think there is something that could have been done regarding the cgo issue. Thanks for tracking down what happened.

@hueypark Thanks for the contribution!

bors r+

petermattis commented:
Unrelated `gossip/restart` failure. But one that deserves to be investigated. I'll file an issue if one doesn't already exist.

bors r+
